File: 173_easy_unitCalculator/173_easy.clj
(use '[clojure.set :only [subset?]])

(defn conversions [n a b]
  (let [unit (filter
               #(subset? (set [a b]) %)
               [{"meters" 1.0
                 "inches" 39.3701
                 "miles" 0.000621371
                 "attoparsecs" 32.4077929}
                {"kilograms" 1.0
                 "pounds" 2.20462
                 "ounces" 35.274
                 "hogsheads of beryllium" 440.7}])]
    (str n " " a
         (if-let [ratio (first unit)]
           (str " is " (* n (/ (ratio b) (ratio a))) " ")
           (str " can't be converted to "))
         b)))

(conversions 23 "pounds" "ounces")
;=> "23 pounds is 368.0008346109534 ounces"
(conversions 3.9 "kilograms" "hogsheads of beryllium")
;=> "3.9 kilograms is 1718.73 hogsheads of beryllium"
(conversions 10 "inches" "miles")
;=> "10 inches is 1.5782814877279968E-4 miles"
(conversions 5 "meters" "ounces")
;=> "5 meters can't be converted to ounces"

File: 173_easy_unitCalculator/173_easy.hs
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE MultiParamTypeClasses #-}

class Fractional v => Unit u v where
    ratio :: u -> v
    convert :: v -> u -> u -> (v, u)
    convert n u1 u2 = (n * (ratio u2 / ratio u1), u2)

data Length = Meters | Inches | Miles | Attoparsecs deriving (Show)
data Mass = Kilograms | Pounds | Ounces | HogsheadsOfBeryllium deriving (Show)
data School = ZoolanderSchoolForKidsWhoCantReadGood | SchoolForAnts deriving (Show)

instance Fractional a => Unit Length a where
    ratio Meters      = 1.0
    ratio Inches      = 39.3701
    ratio Miles       = 0.000621371
    ratio Attoparsecs = 32.4077929

instance Fractional a => Unit Mass a where
    ratio Kilograms            = 1.0
    ratio Pounds               = 2.20462
    ratio Ounces               = 35.274
    ratio HogsheadsOfBeryllium = 440.7

instance Fractional a => Unit School a where
    ratio ZoolanderSchoolForKidsWhoCantReadGood = 1.0
    ratio SchoolForAnts                         = 3.0

